### PR TITLE
Shadows aliasing

### PR DESCRIPTION
**Pull request by lilleyse**
For #2594 
Merge #3665 first

I added a few improvements to fix projective aliasing:
* Render back faces instead of front faces
* Use `polygonOffset`
* Check if the normal is facing away from light (get some soft shading on the edges this way)
    * This is slightly error-prone if the glTF model does not have a varying called `v_normal`, it will still render but the back face shadowing will not work.

For testing purposes I added a tower model from http://www.blendswap.com/blends/view/61653.

### CODE REVIEW

**Comment by lilleyse** [comment]
capture
capture2

**Comment by lilleyse** [comment]
For directly above light source:

Before:
capture4

After:
capture3

**Comment by pjcozzi** [comment]
Is the default shadow map 1K?  The UI says 256, but the aliasing doesn't change when I change it to 1K.

**Comment by lilleyse** [comment]
Yeah it's 1024 but says 256. I'm doing some more updates to the UI and I'll be fixing that.

**Comment by pjcozzi** [comment]
I don't know if it is an issue in this branch or just our glTF loader in general, but the texture doesn't always load.

image

Submit a bug for later if you want.

**Comment by pjcozzi** [comment]
> Render back faces instead of front faces

Just a note that this will not work for all of our geometry types - walls in particular.  The Geometry and Appearances Sandcastle example shows most of the cases.

**Comment by pjcozzi** [comment]
> Apps/SampleData/models/*

Move each new model to a sub-directory.

File: Source/Scene/Model.js [hunk 1]
```diff
@@ -1512,9 +1513,9 @@ define([
         }
 
         // Modify draw program to receive shadows
-        if (frameState.shadowsEnabled && model.receiveShadows) {
+        if (shadowsEnabled && model.receiveShadows) {
             drawVS = ShadowMapShader.createShadowReceiveVertexShader(drawVS);
-            drawFS = ShadowMapShader.createShadowReceiveFragmentShader(drawFS, context);
+            drawFS = ShadowMapShader.createShadowReceiveFragmentShader(drawFS, frameState, 'v_normal');
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Maybe put a TODO here that this is a bit hacky and does not use the glTF attribute semantics.

**Comment by pjcozzi** [comment]
> For testing purposes I added a tower model from http://www.blendswap.com/blends/view/61653.

This is OK.  Please make sure it is as small as possible and credit it (and any changes you made) at the very bottom of LICENSE.md: http://creativecommons.org/licenses/by/3.0/

File: Source/Scene/Model.js (hunk 1 above)
Comment (lilleyse) [inline Source/Scene/Model.js]: glTF doesn't have semantics for varyings though, right?

Comment (pjcozzi) [inline Source/Scene/Model.js]: Ah right.  Hmm, I guess we can only do this reliably when we generate the shader client-side.  Still put a TODO here so we can come up with the most robust approach.

**Comment by lilleyse** [comment]
> I don't know if it is an issue in this branch or just our glTF loader in general, but the texture doesn't always load.

It is working now that I removed references to the normal map.

**Comment by lilleyse** [comment]
> Is the default shadow map 1K? The UI says 256, but the aliasing doesn't change when I change it to 1K.

This is in my next PR, which is mainly just small UI updates.

**Pull request merged**